File: 394-decode-string/decode-string.cpp
class Solution {
public:
    string decodeString(string s) {
        int n=s.size();
        stack<char>st;
        for(int i=0;i<n;++i){
            if(s[i]!=']'){
                st.push(s[i]);
            }
            else {
                string curr_str="";
                while(st.top()!='['){
                    curr_str=st.top()+curr_str;
                    st.pop();
                }
                st.pop();
                string num=""; //because number could be of more than 1 digit
                while(!st.empty() && isdigit(st.top())){
                    num=st.top()+num;
                    st.pop();
                }
                int k=stoi(num);
                while(k--){
                    for(int i=0;i<curr_str.size();++i){
                        st.push(curr_str[i]);
                    }
                }
            }
        }

        string res="";
        while(!st.empty()){
            res=st.top()+res;
            st.pop();
        }
        return res;

    }
};

File: 394-decode-string/decode-string.java
class Solution {
    public String decodeString(String s) {
       int n=s.length();
       if(n==0) return "";
       Stack<Character>st=new Stack<>();
       for(int i=0;i<n;++i){
        if(s.charAt(i)!=']'){
            st.push(s.charAt(i));
        }
        else{
            String currStr=""; 
            while(st.peek()!='['){
                currStr=st.pop()+currStr; //prepare string
            }
            st.pop(); // remove open bracket
            String currNum=""; //number could be greator than 10
            while(!st.empty() && Character.isDigit(st.peek())){
                currNum=st.pop()+currNum;
            }
            int currCount=Integer.parseInt(currNum);
            while(currCount-->0){
                for(int j=0;j<currStr.length();++j){
                    st.push(currStr.charAt(j));
                }
            }
        }
       }
       String resString="";
       while(!st.empty()){
        resString=st.pop()+resString;
       }
       return resString;
    }
}